reconcile/ocm_machine_pools.py

```diff
@@ -1,137 +1,400 @@
+                        cluster=cluster.name,
+                    )
+                )
 
-    return ocm_map, current_state
+    return current_state
 
 
-def fetch_desired_state(clusters):
-    desired_state = []
+def create_desired_state_from_gql(
+    clusters: Iterable[ClusterV1],
+) -> DesiredStateList:
+    desired_state: DesiredStateList = DesiredStateList(cluster_pools=[])
     for cluster in clusters:
-        cluster_name = cluster["name"]
-        machine_pools = cluster["machinePools"]
-        for machine_pool in machine_pools:
-            machine_pool["cluster"] = cluster_name
-            labels = machine_pool.pop("labels")
-            if labels:
-                machine_pool["labels"] = json.loads(labels)
-            taints = machine_pool.pop("taints")
-            if taints:
-                machine_pool["taints"] = taints
-            desired_state.append(machine_pool)
+        if cluster.machine_pools:
+            is_hypershift = False
+            if cluster.spec:
+                is_hypershift = True if cluster.spec.hypershift else False
+            desired_state.cluster_pools.append(
+                DesiredMachinePool(
+                    cluster_name=cluster.name,
+                    hypershift=is_hypershift,
+                    pools=cluster.machine_pools,
+                )
+            )
 
     return desired_state
 
 
-def calculate_diff(current_state, desired_state):
-    diffs = []
-    err = False
-    for d in desired_state:
-        c = [
-            c
-            for c in current_state
-            if d["cluster"] == c["cluster"] and d["id"] == c["id"]
-        ]
-        if not c:
-            d["action"] = "create"
-            diffs.append(d)
-            continue
-        if len(c) != 1:
-            logging.error(f"duplicate id found in {d['cluster']}")
-            err = True
-            continue
-        c = c[0]
-        if c == d:
-            continue
-        if d["instance_type"] != c["instance_type"]:
-            logging.error(
-                f"can not update instance type for existing "
-                f"machine pool {d['id']} in {d['cluster']}"
-            )
-            err = True
-            continue
-        d.pop("instance_type")
-        for key in ["labels", "taints"]:
-            if c.get(key, None) != d.get(key, None):
-                # https://github.com/openshift/machine-api-operator/blob/master/FAQ.md
-                logging.warning(
-                    f"update {key} for machine pool {d['id']} "
-                    f"will only be applied to new Nodes"
-                )
-        d["action"] = "update"
-        diffs.append(d)
+def calculate_diff(
+    current_state: Mapping[str, list[AbstractPool]],
+    desired_state: DesiredStateList,
+) -> tuple[list[PoolHandler], list[InvalidUpdateError]]:
+    diffs: list[PoolHandler] = []
+    errors: list[InvalidUpdateError] = []
+
+    all_desired_pools: list[tuple[str, str]] = []
```

Comment:
Since `all_desired_pools` is only used in contains check, can change to `set[tuple[str, str]]` to improve performance for large dataset.

If you are interested, can also use new generic [differ](https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/utils/differ.py) to simplify diff logic.

Reply:
Changed to a set, good spot!